### PR TITLE
#321 Implemented RqForm.Fake allowing parameters passed in constructor

### LINKED ISSUE #321: RqFake should allow form params
When testing a form processing `Take` there is no clean way of adding form params (URL or body) to the request, one has to do manual encoding and creation of the appropriate string.
Please add such processing to `RqFake` (or maybe other class if appropriate), and describe it in the README section regarding unit tests.

### PR DESCRIPTION
**Pull request by mbarbieri**
Implemented RqForm.Fake allowing parameters passed in constructor.
Added two tests to verify it works and it throws an exception when passing the wrong number of parameters.

Fixes #321

### CODE REVIEW

**Comment by davvd** [comment]
@mbarbieri Thanks for your pull request, let me find someone who can review it

**Comment by davvd** [comment]
@caarlos0 this pull request is for you, please review

File: src/main/java/org/takes/rq/RqForm.java [hunk 1]
```diff
@@ -165,7 +166,6 @@ private static String decode(final CharSequence txt) {
      * Smart decorator, with extra features.
      *
      * <p>The class is immutable and thread-safe.
```
Comment (caarlos0) [inline src/main/java/org/takes/rq/RqForm.java]: @mbarbieri Ler's also close this `<p>` tag

**Comment by caarlos0** [comment]
@mbarbieri I think this "magic" with the array of string is strange... wouldn't it be better to pass a `Map<String,String>` instead?

File: src/main/java/org/takes/rq/RqForm.java (hunk 1 above)
Comment (mbarbieri) [inline src/main/java/org/takes/rq/RqForm.java]: @caarlos0 it's not my code :) anyway, ok, will fix that

**Comment by mbarbieri** [comment]
@caarlos0 it was specifically requested like this. Read the issue #321

File: src/main/java/org/takes/rq/RqForm.java (hunk 1 above)
Comment (caarlos0) [inline src/main/java/org/takes/rq/RqForm.java]: @mbarbieri yeah, I know :)

File: src/main/java/org/takes/rq/RqForm.java [hunk 2]
```diff
@@ -233,4 +233,53 @@ public String single(final CharSequence name, final String def) {
...
+            }
+            this.map = new ConcurrentHashMap<String, List<String>>(0);
+            for (int idx = 0; idx < params.length; idx += 2) {
+                final String key = params[idx];
+                final String value = params[idx + 1];
+                if (this.map.containsKey(key)) {
+                    this.map.get(key).add(value);
+                } else {
+                    final List<String> values = new ArrayList<String>(1);
+                    values.add(value);
```
Comment (caarlos0) [inline src/main/java/org/takes/rq/RqForm.java]: @mbarbieri you can simply use `this.map.put(key, Arrays.asList(value))` here

File: src/main/java/org/takes/rq/RqForm.java [hunk 3]
```diff
@@ -233,4 +233,53 @@ public String single(final CharSequence name, final String def) {
...
+                throw new IllegalArgumentException(
+                    "Wrong number of parameters"
+                );
+            }
+            this.map = new ConcurrentHashMap<String, List<String>>(0);
+            for (int idx = 0; idx < params.length; idx += 2) {
+                final String key = params[idx];
+                final String value = params[idx + 1];
+                if (this.map.containsKey(key)) {
+                    this.map.get(key).add(value);
```
Comment (caarlos0) [inline src/main/java/org/takes/rq/RqForm.java]: @mbarbieri so, if I do:

```java
new RqForm.Fake(
  new RqFake(),
  "foo", "value-1",
  "bar", "value-2",
  "foo", "value-3"
)
```

the `foo` param will have both `value-1` and `value-3`, right?

I don't know, for me it seems "more correct" that the last one overrides the first one... what do you think?

Comment (caarlos0) [inline src/main/java/org/takes/rq/RqForm.java]: @mbarbieri forget about my previous comment, just read again the issue and saw it was requested like this.. sorry

**Comment by caarlos0** [comment]
@mbarbieri just one comment above..

File: src/main/java/org/takes/rq/RqForm.java (hunk 2 above)
Comment (mbarbieri) [inline src/main/java/org/takes/rq/RqForm.java]: Actually I can't... Read [here](https://stackoverflow.com/questions/2965747/why-i-get-unsupportedoperationexception-when-trying-to-remove-from-the-list).
But I changed to something better anyway.

**Comment by mbarbieri** [comment]
@caarlos0 done

File: src/main/java/org/takes/rq/RqForm.java (hunk 2 above)
Comment (caarlos0) [inline src/main/java/org/takes/rq/RqForm.java]: @mbarbieri oh, right, because you might add stuff to it later, got it.

**Comment by caarlos0** [comment]
@rultor merge

File: src/main/java/org/takes/rq/RqForm.java [hunk 4]
```diff
@@ -233,4 +234,56 @@ public String single(final CharSequence name, final String def) {
...
+            if (params.length % 2 != 0) {
+                throw new IllegalArgumentException(
+                    "Wrong number of parameters"
+                );
+            }
+            this.map = new ConcurrentHashMap<String, List<String>>(0);
+            for (int idx = 0; idx < params.length; idx += 2) {
+                final String key = params[idx];
+                final String value = params[idx + 1];
+                if (this.map.containsKey(key)) {
```
Comment (yegor256) [inline src/main/java/org/takes/rq/RqForm.java]: @mbarbieri this code can be done shorter:

```
this.map.putIfAbsent(key, new LinkedList())
this.map.get(key).add(value);
```

that's it.

**Comment by yegor256** [comment]
@mbarbieri one comment from me

File: src/main/java/org/takes/rq/RqForm.java (hunk 4 above)
Comment (mbarbieri) [inline src/main/java/org/takes/rq/RqForm.java]: @yegor256 right...

**Comment by mbarbieri** [comment]
@yegor256 done

File: src/test/java/org/takes/rq/RqFormFakeTest.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+ * @version $Id$
+ */
+public class RqFormFakeTest {
+
+    /**
+     * RqForm.Fake can create fake forms with parameters list.
+     * @checkstyle MultipleStringLiteralsCheck (20 lines)
+     */
+    @Test
+    @SuppressWarnings("PMD.AvoidDuplicateLiterals")
```
Comment (yegor256) [inline src/test/java/org/takes/rq/RqFormFakeTest.java]: @mbarbieri see http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#test-data-coupling

File: src/main/java/org/takes/rq/RqForm.java [hunk 6]
```diff
@@ -233,4 +232,48 @@ public String single(final CharSequence name, final String def) {
...
+         * @param params Parameters list
+         */
+        public Fake(final Request req, final String... params) {
+            super(req);
+            if (params.length % 2 != 0) {
+                throw new IllegalArgumentException(
+                    "Wrong number of parameters"
+                );
+            }
+            this.map = new ConcurrentHashMap<String, List<String>>(0);
```
Comment (yegor256) [inline src/main/java/org/takes/rq/RqForm.java]: @mbarbieri don't you think that it's better to create a body by concatenating `key=value` pairs and then encapsulating an instance of `RqForm.Base(new RqWithBody(req, "key=value&key=value.."))`?

**Comment by yegor256** [comment]
@mbarbieri there are few serious design concerns, see above

File: src/main/java/org/takes/rq/RqForm.java (hunk 6 above)
Comment (mbarbieri) [inline src/main/java/org/takes/rq/RqForm.java]: @yegor256 I don't think converting from an array to a string and then parse the string to create a map is good. I would probably create another one accepting a map (the most obvious data structure for parameters) and then refactor `RqWithBody` and `Fake` to use that.

Comment (yegor256) [inline src/main/java/org/takes/rq/RqForm.java]: @mbarbieri the main problem with current approach is that you don't use `RqForm.Base` at all. It's more effective performance wise, but it makes this Fake class completely independent. See how `RqMultipart.Fake` is implemented

Comment (mbarbieri) [inline src/main/java/org/takes/rq/RqForm.java]: @yegor256 I'm not able to find any `RqWithBody` class

File: src/test/java/org/takes/rq/RqFormFakeTest.java (hunk 5 above)
Comment (mbarbieri) [inline src/test/java/org/takes/rq/RqFormFakeTest.java]: @yegor256 done

**Comment by caarlos0** [comment]
@mbarbieri ping

**Comment by caarlos0** [comment]
@davvd seems like this PR was abandoned...

**Comment by mbarbieri** [comment]
@caarlos0 @davvd working on it... Sorry for the delay...

**Comment by mbarbieri** [comment]
@caarlos0 @davvd done... Really sorry for the delay... 
@yegor256 you really like decorators, don't you? :smile:

**Comment by yegor256** [comment]
@mbarbieri sure, decorators are one of the best OOP tools, see http://www.yegor256.com/2015/02/26/composable-decorators.html

File: src/main/java/org/takes/rq/RqForm.java [hunk 7]
```diff
@@ -267,4 +288,64 @@ public String single(final CharSequence name, final String def)
...
+                );
+            }
+            final StringBuilder builder = new StringBuilder();
+            for (int idx = 0; idx < params.length; idx += 2) {
+                final String key = params[idx];
+                final String value = params[idx + 1];
+                builder.append(key);
+                builder.append(RqForm.KV_SEPARATOR);
+                builder.append(value);
+                builder.append(RqForm.PARAM_SEPARATOR);
```
Comment (caarlos0) [inline src/main/java/org/takes/rq/RqForm.java]: @mbarbieri you can inline key and value and chain the append calls...

**Comment by mbarbieri** [comment]
@davvd really sorry for the delay... I have the latest CR changes already done... it's fine if I won't get payed, can I push it anyway?

**Comment by yegor256** [comment]
@mbarbieri sure, go ahead

**Comment by mbarbieri** [comment]
@caarlos0 @yegor256 done.

**Comment by caarlos0** [comment]
@mbarbieri looks good

File: src/main/java/org/takes/rq/RqForm.java [hunk 8]
```diff
@@ -55,17 +55,25 @@
...
  * @see org.takes.rq.RqGreedy
+ * @since 0.9
  */
 @SuppressWarnings("PMD.TooManyMethods")
 public interface RqForm extends Request {
 
     /**
+     * The parameter separator.
+     */
+    String PARAM_SEPARATOR = "&";
```
Comment (yegor256) [inline src/main/java/org/takes/rq/RqForm.java]: @mbarbieri I don't think our users should see these constants... they should be declared inside classes, as private. not here

File: src/test/java/org/takes/rq/RqFormFakeTest.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    /**
+     * RqForm.Fake can create fake forms with parameters list.
+     * @throws IOException If something goes wrong.
+     */
+    @Test
+    public final void createsFakeWithParams() throws IOException {
+        final String key = "key";
+        final String akey = "anotherkey";
+        final String value = "value";
+        final String avalue = "anothervalue";
```
Comment (yegor256) [inline src/test/java/org/takes/rq/RqFormFakeTest.java]: @mbarbieri try to use `a&b` here instead of `anothervalue` and your test will fail

**Comment by yegor256** [comment]
@mbarbieri you forgot about URL encoding, see my comments above

**Comment by mbarbieri** [comment]
@yegor256 added the URL encoding.

File: src/main/java/org/takes/rq/RqForm.java (hunk 8 above)
Comment (mbarbieri) [inline src/main/java/org/takes/rq/RqForm.java]: @yegor256 sorry, feeling stupid now... but If I declare them inside classes as private Qulice complains about having those strings twice in the same file, if I keep them in the interface they are clearly accessible from the outside... I'm not able to think about anything else... a part from excluding some Qulice check

Comment (yegor256) [inline src/main/java/org/takes/rq/RqForm.java]: @mbarbieri yes, you can suppress Qulice in this case

**Comment by caarlos0** [comment]
@yegor256 @mbarbieri ping

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by caarlos0** [comment]
@yegor256 ping

**Comment by yegor256** [comment]
@caarlos0 why pinging me? I'm not the author of this branch :)

**Comment by caarlos0** [comment]
@yegor256 oops, sorry

**Comment by caarlos0** [comment]
@mbarbieri  ping

**Comment by caarlos0** [comment]
@davvd seems like this was abandoned...

**Comment by yegor256** [comment]
@mbarbieri I'm closing it as there is no more activity...

**Comment by mbarbieri** [comment]
@yegor256 @caarlos0 sorry it was abandoned because someone else picked up the ticket and I didn't close this PR...

**Comment by davvd** [comment]
> @davvd seems like this was abandoned...

@caarlos0 yes, agreed, @yegor256 will close it

**Comment by davvd** [comment]
@elenavolokhova please, review this ticket for compliance with our [QA rules](http://www.teamed.io/qa.html)

**Comment by elenavolokhova** [comment]
@caarlos0 
According to our [quality rules](http://www.teamed.io/qa.html):
 > Comments were mostly about design problems, not cosmetic issues.
 
 There were mostly cosmetic issues found during review. Try to pay attention to bigger design problems and the overall quality of the code.
Please confirm that this new rule is clear.

**Comment by caarlos0** [comment]
@elenavolokhova clear

**Comment by elenavolokhova** [comment]
@davvd Quality is acceptable here.

**Comment by elenavolokhova** [comment]
@davvd Quality is acceptable here

**Comment by davvd** [comment]
> @davvd Quality is acceptable here.

@elenavolokhova thank you, let's try to make it "good" next time

**Comment by davvd** [comment]
@caarlos0 I added 10 mins to @elenavolokhova (for QA review) in transaction 62640307

**27 mins** added to your account (payment number `AP-1A4902663B529702P`), many thanks for your contribution! 1493 hours and 11 mins spent here.

review comments (c=12) added as a bonus

+27 to your rating, your total score is [+1690](http://www.netbout.com/b/35095?open=rating)

**Pull request closed without merging**